airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java

```diff
@@ -82,7 +108,194 @@ public JsonNode toDatabaseConfig(JsonNode mssqlConfig) {
         "spt_values",
         "spt_fallback_usg",
         "MSreplication_options",
-        "spt_fallback_dev");
+        "spt_fallback_dev",
+        "cdc"); // is this actually ok? what if the user wants cdc schema for some reason?
+  }
+
+  @Override
+  public AirbyteCatalog discover(JsonNode config) throws Exception {
+    AirbyteCatalog catalog = super.discover(config);
+
+    if (isCdc(config)) {
+      final List<AirbyteStream> streams = catalog.getStreams().stream()
+          .map(MssqlSource::removeIncrementalWithoutPk)
+          .map(MssqlSource::setIncrementalToSourceDefined)
+          .map(MssqlSource::addCdcMetadataColumns)
+          .collect(toList());
+
+      catalog.setStreams(streams);
+    }
+
+    return catalog;
+  }
+
+  @Override
+  public List<CheckedConsumer<JdbcDatabase, Exception>> getCheckOperations(JsonNode config) throws Exception {
+    final List<CheckedConsumer<JdbcDatabase, Exception>> checkOperations = new ArrayList<>(super.getCheckOperations(config));
+
+    if (isCdc(config)) {
+
+      // note, it's possible these queries could fail if user doesn't have correct permissions
+      // hopefully in these cases it should be evident from the SQLServerException thrown
+
+      // check that cdc is enabled on database
+      checkOperations.add(database -> {
```

Comment:
I would suggest refactoring this part to define these check operations outside of `getCheckOperations` then reuse them inside it. For example: 

```

  protected void assertCdcEnabledInDb(JsonNode config, JdbcDatabase database) throws SQLException {
    List<JsonNode> queryResponse = database.query(connection -> {
      final String sql = "SELECT name, is_cdc_enabled FROM sys.databases WHERE name = ?";
      PreparedStatement ps = connection.prepareStatement(sql);
      ps.setString(1, config.get("database").asText());
      LOGGER.info(String.format("Checking that cdc is enabled on database '%s' using the query: '%s'",
          config.get("database").asText(), sql));
      return ps;
    }, JdbcUtils::rowToJson).collect(toList());

    if (queryResponse.size() < 1) {
      throw new RuntimeException(String.format(
          "Couldn't find '%s' in sys.databases table. Please check the spelling and that the user has relevant permissions (see docs).",
          config.get("database").asText()));
    }

    if (!(queryResponse.get(0).get("is_cdc_enabled").asBoolean())) {
      throw new RuntimeException(String.format(
          "Detected that CDC is not enabled for database '%s'. Please check the documentation on how to enable CDC on MS SQL Server.",
          config.get("database").asText()));
    }
  }
  
  protected void assertCdcSchemaQueryable(JsonNode config, JdbcDatabase database) throws SQLException {
    List<JsonNode> queryResponse = database.query(connection -> {
      final String sql = "SELECT * FROM cdc.change_tables";
      PreparedStatement ps = connection.prepareStatement(sql);
      LOGGER.info(String.format("Checking user '%s' can query the cdc schema and that we have at least 1 cdc enabled table using the query: '%s'",
          config.get("username").asText(), sql));
      return ps;
    }, JdbcUtils::rowToJson).collect(toList());

    if (queryResponse.size() < 1) {
      throw new RuntimeException("No cdc-enabled tables found. Please check the documentation on how to enable CDC on MS SQL Server.");
    }
  }
```

Then we can use them as follows: 

```
if (isCdc(config)) {
      checkOperations.add(database -> assertCdcEnabledInDb(config, database));
      checkOperations.add(database -> assertCdcSchemaQueryable(config, database));
...
```

this makes it more readable and makes it possible to unit test these helper methods. Some good tests to have would be: 

1. Verify that each method does what it's supposed to e.g: if cdc is not enabled in the Db then throws exception etc.
2. that `getCheckOperations` returns these operations when cdc is enabled

to do no. 2 it might be easier to define each of these as a class which implements `CheckedConsumer`

Reply:
Addressed: https://github.com/airbytehq/airbyte/pull/4689/commits/2f87d62186ad5bd99b496afe4a0529b9b0e099cb